### ISSUE TITLE
server: decommission->decommissioned transition causes abrupt loss of service, need to introduce final wait

### ISSUE THREAD

**Issue by smcvey** [cockroachdb/cockroach#72754]
**Describe the problem**

~A decommissioned node is supposed to still be able to serve queries from connected clients to the rest of the cluster.~

The result of an orchestration signal to take a node out should always ensure that clients are redirected away (via LB redirect, then gracefull client conn shutdown). This is the role of the interaction between the readiness probe, the SQL connection shutdown etc that are part of the drain process.

However, since 21.1, the **decommission** process does not gracefully let clients out. 

In 21.1.0, for example, queries just hang.  In 21.1.11, clients receive an error message like this:

```
root@:26257/defaultdb> select * from x;
ERROR: failed to connect to n2 at 172.28.128.25:26257: initial connection heartbeat failed: rpc error: code = PermissionDenied desc = n5 was permanently removed from the cluster at 2021-11-15 15:52:19.810375648 +0000 UTC; it is not allowed to rejoin the cluster
```

**To Reproduce**

Create a 5 node cluster, then decommission node 5.  In the meantime, connect to node 5 using `cockroach sql` and run some simple requests such as `SHOW DATABASES;`.  This will continue to work fine until the final range has migrated away from the node and the node is removed from the cluster.  Then the error messages start appearing for each request.

**Expected behavior**
A decommissioned node should continue to serve all requests just like it could in 20.2 and earlier.

**Environment:**
 - CockroachDB version: 21.1.11 (possibly 21.2.0 - untested)

**Additional context**
Caused error messages to the connected application.

gz#10501

Epic crdb-11843

Epic CRDB-11843

**Comment by knz** [comment]
cc @piyush-singh for triage

**Comment by knz** [comment]
clarification about the title:  given  the node won't be able to serve requests after put in the "decommissioned" state, it needs to remove itself from liveness and LBs **and then wait a bit** before that final transition, like is done for a regular drain already.

**Comment by smcvey** [comment]
Tested on 21.2.0 which dropped today.  Also broken.

**Comment by rbkrabbe** [comment]
> remove itself from liveness and LBs **and then wait a bit** before that final transition

It would be nice if the wait could be whatever is specified in the `server.shutdown.drain_wait` cluster setting as this is already used to allow LBs to remove the endpoint.

**Comment by smcvey** [comment]
I've been testing and in 20.2, a decommission doesn't close or disconnect any clients.  So the `server.shutdown.drain_wait` setting has no effect - this is 'good behaviour' and expected.

Where `server.shutdown.drain_wait` is used is when, after decommissioning the node, the operator then sends a SIGTERM signal or uses `cockroach node drain` - the connected session is still able to run queries until `server.shutdown.drain_wait` has expired.  This behaviour must happen again once this issue is resolved.  I suspect there hasn't been any change to this behaviour, but obviously it's not possible to tell with the current bug in place.

**Comment by irfansharif** [comment]
Found this issue through an internal slack thread.

> A decommissioned node should continue to serve all requests just like it could in 20.2 and earlier.

I'm having trouble understanding the issue, what the ask is, and whether this is a bug. Why should decommissioned nodes continue to serve requests? They can't, they're no longer part of the cluster -- a guarantee we rely on in other parts of the system. When upgrading a cluster for e.g. we need to ensure that all current members of the cluster are running some the new binary and then executed some migration triggers before declaring the upgrade process complete. Given that these nodes can't serve requests (by design!), what is this issue asking for? Perhaps this is what @knz is alluding to above with node draining, but I'm not following.

**Comment by knz** [comment]
@irfansharif it's important to ensure continuity of service, that load balancers learn of the impending node unavailability before it happens.

When a user *shuts down a node* (using the drain process), they benefit from the careful dance we implemented there, to ensure e.g. that:

1) the readiness probe says "no" for a configurable long while, so LBs know to look somewhere else
2) the thing actively closes SQL connections, but not before waiting for some time after starting to refuse opening new sessions
3) some more delay logic that the SQL team is currently building

With the decommissioning process, none of that careful orchestration is happening: a node appears to be online until poof! it's not online any more, with no warning. All the SQL client conns open at that point become unusable, and there's loss of service without advance notice.

That's the unacceptable regression.

Before the decommission refactor, a decommissioning node would remain decommissioning *and continue to redirect traffic to other nodes*, until it would be shut down gracefully, at which point the careful drain dance described above would kick in, and all was well.

After the refactor, the node becomes unusable before the graceful shutdown is initated. This is not nice / not desirable.

We must ensure that the careful dance to let clients out starts *and completes* before the node is switched to the new "decommissioned" state.

**Comment by piyush-singh** [comment]
I believe (and Raphael can probably speak better to this) that the intended UX is that decommissioning/decomissioned nodes continue to serve traffic until a `cockroach node drain` is initiated. That final step should mark the node as no longer accepting connections instead of implicitly handling that as part of the decommission step. This behavior changed after 20.2 which has broken automation for some of our customers. See: https://cockroachlabs.slack.com/archives/CQ2THFH1C/p1637223516027500

**Comment by knz** [comment]
> continue to serve traffic until a cockroach node drain is initiated. That final step should mark the node as no longer accepting connections instead of implicitly handling that as part of the decommission step

No that's not what I'm suggesting (the flip to the `decommissioned` state would not allow it)

What I'm suggesting is that the server would enter the node shutdown logic *before it switches itself to "decommissioned"*.

**Comment by piyush-singh** [comment]
I see, thank you for the clarification Raphael

**Comment by irfansharif** [comment]
To see if I understand correctly/concisely: we're asking that before a node is fully decommissioned, it should be drained? If so, can we update the issue text.

**Comment by knz** [comment]
> we're asking that before a node is fully decommissioned, it should be drained?

something like that; more specifically, it should behave as if it it was shutting down.

> if so, can we update the issue text?

for the title, the word "drain" is too ambiguous nowadays and too many people think it means "make the process shut down gracefully".

That is why I used "need to introduce final wait".

**Comment by irfansharif** [comment]
Understood, thanks for explaining. I'll leave it to y'all to track it -- was just sanity checking it didn't fall under KV.

**Comment by knz** [comment]
> was just sanity checking it didn't fall under KV.

It may fall partly under KV: we will want to introduce additional conditions to be satisfied before the flip `decommissioning` -> `decommissioned`.

**Comment by knz** [comment]
Alternatively, the customer can simply:

1. shut down the node gracefully
2. decommission it *after it's been shut down*

This would be reliable. Is there a reason not to do this?

**Comment by knz** [comment]
@data-matt could you check out the last two comments above and see what the customer thinks? Our internal preference goes for the 2nd workaround.

For reference, a node can be decommissioned after it's shut down, using another node by issuing the command `cockroach node decommission N1 --host N2` (N2 healthy, N1 node to decommission)

**Comment by knz** [comment]
This older issue, if we had worked on it sooner, would have helped us detect the situation earlier: #41017

**Comment by knz** [comment]
Customer reports that "shut down, then decommission" is not nice because it makes some ranges underreplicated. This is suggested instead:

1. `node drain` on the node to decommission, to get LB/clients out but leave the node running with all replicas (avoid underreplication)
2. `node decommission` from another node to move replicas away
3. when decommission complete, shut down process

this way we get graceful exit of clients and avoid underreplication

**Comment by irfansharif** [comment]
Just confirmed that draining the to-be-decommissioned node first and then decommissioning it from elsewhere does what we want: it'll drain clients on that node, get the LBs to point elsewhere, start moving off replicas from that node in the background, and once done, we're in the clear to shut the machine down.

Looking through our docs, I feel this could be made clearer. We could also add checks to the decommissioning process to ensure that the node being decommissioned is already drained. That affects `decommission --self`, which would no longer be able to establish a connection with the node seeing as how it's supposed to be drained first. Perhaps we reconsider --self, or log a loud warning about the suggested control flow.

**Comment by irfansharif** [comment]
@piyush-singh, @knz: I'm going to let this issue sit on the server team's plate since it's the closest owner we have and it concerns itself around the UX of the server process lifetime. Like mentioned, a lot of this can be documented a lot more prominently. I fumbled my through too; a single page to collate all of decommissioning/recommissioning/draining, the exact lifecycle of a node, the health checks of interest, etc. would make this much clearer to all of us. 

Another massive UX benefit IMO would be the ability to send a drain request to a remote node. It would be symmetric with the decommission CLI usage.

TL;DR - before decommissioning a node, it's best practice to always drain it first and then decommission the node from another healthy machine.

**Comment by irfansharif** [comment]
Also ensuring that nodes are drained before being decommissioned lets us avoid the very bizzare UX of allowing SQL shells to be established against a decommissioned node, where nothing is actually usable.

**Comment by irfansharif** [comment]
Suggested action items:
- [ ] Add a check in the decommissioning process to ensure the targeted node was already drained.
- [ ] Deprecate (or remove) `decommission --self`, or log loudly against it being an anti-pattern
- [ ] (Nice to have) Add the ability to drain a node from another node
- [ ] Generate a tech-note/doc page to spell out the "lifecycle of the server process"; spell out how decommissioning/recommissioning/draining interoperate. Talk about the observable health checks (/health/ready?=1) and how the relate to the latter.

**Comment by piyush-singh** [comment]
> a single page to collate all of decommissioning/recommissioning/draining, the exact lifecycle of a node, the health checks of interest, etc. would make this much clearer to all of us

@taroface and I are planning to overhaul our docs for this over the course of this quarter. I'll work with Ryan to prioritize this work when he returns next week. 

> We could also add checks to the decommissioning process to ensure that the node being decommissioned is already drained. 

This would make the UX much clearer and could return a useful error that instructs the user on how to proceed. I can file an issue.

>Another massive UX benefit IMO would be the ability to send a drain request to a remote node. It would be symmetric with the decommission CLI usage.

Can you expand on where this would be useful? Just want to make sure I understand your point

**Comment by irfansharif** [comment]
> Can you expand on where this would be useful?

Nothing more than a UX benefit. I've marked it as a "nice to have" above. It's easy enough to do -- today when draining a node, the CLI issues an RPC. We could issue that RPC from another node as well. It will make it symmetric with the decommissioning UX where, like above, I'm proposing we only allow decommissioning from another node. Requiring decommissioning from another node but draining from the same node seems odd to me is all.

**Comment by piyush-singh** [comment]
Ah I see, yes that makes sense. When you said there was a massive benefit, I was worried I missed something else. Thanks for clarifying

**Comment by data-matt** [comment]
Hi @knz/@piyush-singh, we have been informed that the workaround provided by this ticket is unsuitable.

**Comment by knz** [comment]
@data-matt more details? unsuitable how?

**Comment by data-matt** [comment]
From a brief discussion yesterday, it was confirmed that the workaround for the manual decommissioning of a node works however, it seems to be cumbersome for the operators to programmatically handle the node drain and node decommission apis.

**Comment by knz** [comment]
@jtsiros proposes the following acceptance criteria: (I'm purely quoting)

> - Add a check in the decommissioning process to ensure the targeted node was already drained.
> - Add the ability to drain a node from another node
> - Deprecate `decommission --self`, and log loudly against it being an anti-pattern (including the deprecation) and provide clear documentation on the correct node removal procedure (drain then decom)
> - Generate a doc page to spell out the "lifecycle of the server process"; spell out how decommissioning/recommissioning/draining interoperate. Talk about the observable health checks (`/health/ready?=1`) and how they relate to the latter.

**Comment by knz** [comment]
> Add a check in the decommissioning process to ensure the targeted node was already drained.

This one is complicated. Draining is a process, not a state. After a process has been draining, it can shutdown, *then restart* then not be draining any more. 

So what we can check instead is whether a node is *currently down*.

There is an issue remaining with this idea however: *when exactly* do we check this? if we merely check that the target node is down only at the beginning of the decommissioning process, then it could be restarted after the decommissioning process starts, and we'd still have the issue. 

We could improve this perhaps by doing two separate things:
- check when decommissioning starts, that the target node is down
- check when any node starts, that it is not decommissioning already (and if it is, refuse to start)

However, adding this interlock comes with two additional hurdles:

- whether a node is down or not is not determined atomically. So the interlock above is not a sure-proof way to avoid badness. It's possible for a node to be considered down by another node (starting the deco process), while it was actually already restarting but not updated its liveness record already.

- by refusing to start a decommissioning node, we create the risk to deadlock a cluster: by marking all nodes as decommissioning, and stopping the processes accidentally eg via an automated upgrade, we'd become unable to restart the cluster.

So the bottom line is this: IMHO we cannot **enforce the invariant that only down nodes can be decommissioned** safely. What we can do perhaps, is produce an *advisory* error (that can be overridden) in the decommissioning call, when it starts, if the target node is not down. And document heavily that it is dangerous to restart a node that has been marked as decommissioning.

> Add the ability to drain a node from another node

that is likely useful, but likely orthogonal to the customer issue at hand. I'd track this in a different issue.

> Deprecate decommission --self, and log loudly against it being an anti-pattern (including the deprecation) 

agreed

> Generate a doc page to spell out the "lifecycle of the server process"; spell out how decommissioning/recommissioning/draining interoperate

Super useful, but completely orthogonal to the customer issue at hand. I'd track progress on this separately.

**Comment by knz** [comment]
> Add a check in the decommissioning process to ensure the targeted node was already drained.

This one is complicated. Draining is a process, not a state. After a process has been draining, it can shutdown, *then restart* then not be draining any more. 

So what we can check instead is whether a node is *currently down*.

There is an issue remaining with this idea however: *when exactly* do we check this? if we merely check that the target node is down only at the beginning of the decommissioning process, then it could be restarted after the decommissioning process starts, and we'd still have the issue. 

We could improve this perhaps by doing two separate things:
- check when decommissioning starts, that the target node is down
- check when any node starts, that it is not decommissioning already (and if it is, refuse to start)

However, adding this interlock comes with two additional hurdles:

- whether a node is down or not is not determined atomically. So the interlock above is not a sure-proof way to avoid badness. It's possible for a node to be considered down by another node (starting the deco process), while it was actually already restarting but not updated its liveness record already.

- by refusing to start a decommissioning node, we create the risk to deadlock a cluster: by marking all nodes as decommissioning, and stopping the processes accidentally eg via an automated upgrade, we'd become unable to restart the cluster.

So the bottom line is this: IMHO we cannot **enforce the invariant that only down nodes can be decommissioned** safely. What we can do perhaps, is produce an *advisory* error (that can be overridden) in the decommissioning call, when it starts, if the target node is not down. And document heavily that it is dangerous to restart a node that has been marked as decommissioning.

> Add the ability to drain a node from another node

that is likely useful, but likely orthogonal to the customer issue at hand. I'd track this in a different issue.

> Deprecate decommission --self, and log loudly against it being an anti-pattern (including the deprecation) 

agreed

> Generate a doc page to spell out the "lifecycle of the server process"; spell out how decommissioning/recommissioning/draining interoperate

Super useful, but completely orthogonal to the customer issue at hand. I'd track progress on this separately.

**Comment by knz** [comment]
Erik reminds me we also need to address this: #62693

**Comment by knz** [comment]
In #73763, @bdarnell [explains](https://github.com/cockroachdb/cockroach/issues/73763#issuecomment-993689472): 

> Decommissioning is (intended to be) the fastest way to remove a node from the cluster without consuming any of your fault tolerance budget (which is just a single node if you have any RF=3 ranges).  [...] The entire purpose of decommissioning is to have a non-failure way to remove all ranges from a live healthy node. We need to continue to supply that (maybe we decouple the "remove replicas" process from the final "permanent removal" step, but we do need a process that doesn't begin with shutting the node down).

In other words, the idea to change our best practices to require a node to be down before decommissioning starts is a non-starter.

Instead, we should push for the original ask, which is to enhance the deco process to finish with a graceful exit for connected clients. This is a more complex ask, but after Ben's argument more aligned with our advertised recovery objectives.

**Comment by smcvey** [comment]
Question from Lance Lierheimer, CEA:

```
Is there an update on when this will be backported into 21.1? This is a blocker for Cloud Kitchens to deploy 21.1 into production and I need to update them today. Thanks!
```

**Comment by cameronnunez** [comment]
@smcvey the fix is still in progress, so no update yet. In the meantime, there is the [manual decommissioning of a node workaround](https://github.com/cockroachdb/cockroach/issues/72754#issuecomment-973119620).

**Comment by lancel66** [comment]
> Question from Lance Lierheimer, CEA:
> 
> ```
> Is there an update on when this will be backported into 21.1? This is a blocker for CUSTOMER to deploy 21.1 into production and I need to update them today. Thanks!
> ```
Please remove the customer name from the comment. Also, the manual decommission is not sufficient for them to deploy into production. THe regression bug has broken their deployment automation so it is still a blocker. Thanks!

**Comment by jtsiros** [comment]
> Instead, we should push for the original ask, which is to enhance the deco process to finish with a graceful exit for connected clients. This is a more complex ask, but after Ben's argument more aligned with our advertised recovery objectives.

@knz just to summarize (and check my understanding), we are proposing having the node remain in the `decommissioning` state for the duration of a user configurable `final_wait` (or whatever we call it) flag then mark node as `decommissioned` rather than reverting back to what's documented [here](https://www.cockroachlabs.com/docs/stable/remove-nodes.html#step-3-check-the-node-and-cluster-after-the-decommissioning-process)?

> It may fall partly under KV: we will want to introduce additional conditions to be satisfied before the flip decommissioning -> decommissioned.

If this is still the case, can you enumerate those here?

**Comment by knz** [comment]
> we are proposing having the node remain in the decommissioning state for the duration of a user configurable final_wait (or whatever we call it) fla

No, we are proposing to hook up the entire drain logic (i.e. all the work performed by the Drain RPC) at the end of the decommissioning process. Just some delay won't cut it - we need the full process to make it equivalent to a graceful shutdown.

> rather than reverting back to what's documented here?

I don't understand this part

> >  It may fall partly under KV: we will want to introduce additional conditions to be satisfied before the flip decommissioning -> decommissioned.
>
> If this is still the case, can you enumerate those here?

Cameron and I discussed that yesterday; what I mean is that I don't personally know where in the source code that state is being changed, and I fear it's somewhere deep in KV land. We need to inject a callback in there, so that the callback can do the drain work. I don't (yet) know how to do this and we need someone in KV land to help.

**Comment by knz** [comment]
@irfansharif has contributed some additional thoughts, which I'll try to summarize here: instead of waiting until the end of the decommissioning process to start draining clients, it's probably beneficial to start marking the node as non-ready (but not non-healthy, and not non-live) at the beginning of the decommission. This way, SQL load can start moving off the decommissioning node.

However, I (@knz) think this runs amok of @bdarnell recommendation to think about the cluster's ability to tolerate a node failure. If there are, say, 3 nodes; and we start decommissioning one of them, at that point all 3 are able to get SQL traffic. If we start refusing SQL client conns (by redirecting the LB away) as soon as 1 of them starts decommissioning, then the other 2 nodes are taking each 1/6 extra traffic. If then one of the 2 remaining node dies, then the one remaining node suddenly needs to support the entire cluster load. That's not safe.

**Comment by knz** [comment]
I somewhat sense we have contradictory design goals here, and I'd prefer to have @bdarnell and a PM in the room before we commit to an implementation course:

- do we want to continue serving SQL queries during the decommissioning, and only stop at the end after all replicas have moved away?
- or, do we prefer to move the SQL traffic off the node at the beginning of the decommissioning process?

How to think about this properly? I will admit I am not comfortable reasoning about recovery objectives.

**Comment by irfansharif** [comment]
Thanks for summarizing Rafa. @tbg, your input here would be valuable as well. In KV's past escalations and the general theme of "decommissioning causes disruptions in end-user latencies" can be attributed to us trying to serve SQL queries while a node is being decommissioned (the first option in https://github.com/cockroachdb/cockroach/issues/72754#issuecomment-996813993). This is something we've been asked to try and smoothen out the UX for -- we can of course suggest users to drain the to-be-decommissioned nodes first to reduce this disruption, and some in fact do, but I question what our default behaviour is.

> do we prefer to move the SQL traffic off the node at the beginning of the decommissioning process?

I can see two downsides in this approach, and I'll try to argue that they're good tradeoffs to make:
(a) While a node (N1) is being decommed+drained, nodes containing copies of N1 will experience additional traffic (that slowly declines) while data is shuffled out of N1 out to the rest of the cluster.
(b) If another node with data fails while N1 is being decommed+drained, we'll end up with a single replica being able to serve SQL requests. We're not losing quorum! Just N1 is unable to serve requests. At this point operators would either have to wait for the all the data on N1 to have been bussed out of it, or cancel the decomm+drain process (maybe even "undraining" N1, to let it start serving SQL requests for the data it still has).

Re: (a), this seems fine to me, healthy clusters should not be so underprovisioned such that a replica unable to serve SQL requests causes other nodes to tip over. As for (b), that too seems fine?

**Comment by knz** [comment]
> We're not losing quorum! 

That remains to be seen. If 1 out of 3 node suddently must shoulder 3x more traffic, chances are it will fall over and die. Then we really lose quorum.

**Comment by irfansharif** [comment]
If we want to implement both variants (the options differ only in whether the node being decommissioned is drained first), perhaps a flag is one compromise. I would still argue the defaults here should be just draining it anyway -- I'll point to our accumulated anecdata of basically never encountering quorum loss scenarios except as a result of operator fat fingering. The 1/3 nodes shouldering 300% its baseline traffic is in the worst case where we're only running a 3 node cluster -- in this situation we (a) can't decommission anything anyway[*], and (b) can be brought back down to 150% by canceling the inflight decomm/draining process. This is something we can do automatically by the way, by looking at liveness info and bailing if there are any _additional_ nodes that have gone during the decomm process.

[*] If we go with draining first before decommissioning where decommissioning is "stuck" because there are no valid upreplication targets (zone configs are a frequent offendor), maybe that's undesirable?

**Comment by tbg** [comment]
On a technical level my opinion is that it is prudent to leave decommissioniong alone and to tell operators to either drain or otherwise shift traffic away from the nodes they intend to decommission before this process starts, and to focus technical efforts on making this happen (checks, docs).

The question then is if that is acceptable to the PMs that are in charge of our decommissioning workflows, i.e. if after doc updates, etc, are our customers happy with this (or should they be)? We would probably suggest to decommission only draining nodes (if they're down, that's a moot point of course). Perhaps the decommission cli could warn (after a grace release, fail until --force flag provided?) when a destination node is live and not draining. Users whose conn pool is not tied to CRDB could disregard the warning.

As a historical datapoint, we used to mark nodes as draining at the beginning of decommissioning, and this too produced lots of undesirable behavior for our customers who then tended to accidentally start decommissioning more nodes than they wanted, and tanked their clusters. We switched away from that to the current model in which decommissioning is mostly just a hint for the allocator with a user hook to learn when things have moved off. The recent long-running migrations safety work now requires additional care because the cluster will be hard-rejected from the cluster. I still think we shouldn't add additional bells and whistles to the decommissioning process as there's a good chance we'll find those cause problems elsewhere later. Maybe going back to marking these nodes as draining is okay now, but I would be very deliberate about it if that's what we want to do, as it's an invasive change.

We could probably work on the UX for SQL connections to decommissioned nodes (that are experiencing lock-out from the cluster), i.e. fail them quickly. I'm also not sure what our health endpoint will do for such a decomissioned node; I would assume it should turn red as it won't be able to maintain liveness. But this is just putting lipstick on a pig, queries should no longer be served by such nodes when they get excluded from the cluster, so the best thing I can see is making sure our customers understand that and only decommission a live non-draining node when this is really what they want (& which is not the default).

**Comment by knz** [comment]
tobias, I'll reply to your comment below.

in response to irfan above:

if an operator specifically wants to drain at the start of decommission, they can do this already. They can run the command `cockroach node drain` . If they then suddenly need to "undrain" they can restart the node process, which will bring it back to an undrained, decommissioning state.

I had a brief chat with @mwang1026 about this; michael forced me to write a tldr for the situation:

> - CK is unhappy about how decommissioning is disruptive
> - irfan says "let's move SQL away as soon as decommission starts"
> - ben says "let's move SQL away at the end of decommission, just beofre the node becomes unavailable"
>  and what is different between what we did in 20.2 vs 21.1 that is causing issues?
> - in v20.2, we used to move SQL away at the end of decommission
> any context on why we changed that in 21.1?
> - it was an oversight

I think the customer would like us to do something similar to what we had in v20.2, which is drain at the end of decommissioning. This is the backward-compatible option. Ben's opinion is congruent with that. Moreover, if a user wants to drain at the beginning, they have options to do so.

**Comment by knz** [comment]
> On a technical level my opinion is that it is prudent to leave decommissioniong alone and to tell operators to either drain or otherwise shift traffic away from the nodes they intend to decommission before this process starts, and to focus technical efforts on making this happen (checks, docs).

We started this entire issue thread under this premise (after a discussion on slack that was suggesting this). However, then Ben reminded us [here](https://github.com/cockroachdb/cockroach/issues/73763#issuecomment-993689472) that it's valuable to have a decommissioning node (emphasis: -ing) participate in supporting query load, to help with the ability of the cluster to tolerate another node fault.

This is why we are revisiting this goal now.

> The question then is if that is acceptable to the PMs that are in charge of our decommissioning workflows [...]

We have proposed that option and the particular customer who is pushing us here is telling us "no".

> As a historical datapoint, we used to mark nodes as draining at the beginning of decommissioning, and this too produced lots of undesirable behavior for our customers who then tended to accidentally start decommissioning more nodes than they wanted, and tanked their clusters. We switched away from that to the current model in which decommissioning is mostly just a hint for the allocator with a user hook to learn when things have moved off.

Yes, folk are comfortable with this option. The one thing that doesn't work in the current approach is that the transition at the end of the decommissioning process (emphasis: -ing) to the decommissioned (emph: -ed) state is too abrupt; SQL client sshould be moved off (using the regular draining dance) gracefully before this transition takes place.

> The recent long-running migrations safety work now requires additional care because the cluster will be hard-rejected from the cluster. I still think we shouldn't add additional bells and whistles to the decommissioning process as there's a good chance we'll find those cause problems elsewhere later.

What kind of problem do you foresee with adding a regular graceful client drain process at the end of the decommissioning process, before the transition to decommissioned?

>  Maybe going back to marking these nodes as draining is okay now, but I would be very deliberate about it if that's what we want to do, as it's an invasive change.

That is not what is being discussed here.

> We could probably work on the UX for SQL connections to decommissioned nodes (that are experiencing lock-out from the cluster), i.e. fail them quickly. 

That is also not what is being discussed here. Everything that happens after the node has been marked as decommissioned (emph: -ed) is too late. We're discussing what should happen **before** the transition to decommissioned (emph: -ed) takes place.

**Comment by tbg** [comment]
> What kind of problem do you foresee with adding a regular graceful client drain process at the end of the decommissioning process, before the transition to decommissioned?

That seems fairly reasonable. You envision doing this in the `cli`, right? Or on the server side (so we'd need a way for node X to drain node Y and wait for this to complete). We also probably need some auto-detection logic that decides whether the node is even there (if it's offline, no point in draining it). So probably a `--drain=(auto|off|on)` flag where `on` requires drain (will fail if node not reachable), `auto` drains if node reports to liveness/is reachable via drain RPC, and `off` is today's behavior. If some variation of this is what the proposal is - SGTM and, independently, thanks for explaining

**Comment by knz** [comment]
> You envision doing this in the cli, right? Or on the server side (so we'd need a way for node X to drain node Y and wait for this to complete).

The latter I think, but if the cost of doing so is too great maybe we can fall back to the former which I know is cheaper.

> We also probably need some auto-detection logic that decides whether the node is even there (if it's offline, no point in draining it). So probably a --drain=(auto|off|on) flag where on requires drain (will fail if node not reachable), auto drains if node reports to liveness/is reachable via drain RPC, and off is today's behavior.

I think by proposing this you're already technically a mile ahead of everyone else in this convo. I don't disagree with you, but I myself don't know yet what is best.

I'm still stuck at the question "do we push SQL clients at the beginning or at the end" since Irfan and Ben seem to have differing opinions on the matter.

**Comment by irfansharif** [comment]
Actually I ~may have~ confused myself. I think _automatically_ draining at the end of decommissioning is fine, I just want to make sure that we do drain at some point. For our default recommendations, I would recommend operators to manually drain the to-be-decommissioned nodes first before decommissioning, and updating our CLI through warnings or opt-in flags to bypass the "is the node that you want to decommission already drained/not live" check. https://github.com/cockroachdb/cockroach/issues/72754#issuecomment-973307143.

**Comment by irfansharif** [comment]
Given all the confusing back and forth (thanks in part to yours truly), it would be clearer to everyone to abandon this issue in favor of another with more distilled recommendations and next steps.

**Comment by knz** [comment]
I'm ok continuing to use this issue, as long as we do periodic rollups with the knowledge and decisions so far.

So in summary so far:

- Customer and Irfan say we want to drain SQL clients away _at some point_ (instead of not at all, as currently implemented)
- Ben says this should happen at the end
- Tobias and I don't have a preference about whether it should be at the beginning, during or at the end
- I think that from an implementation perspective "at the end" is easier and less likely to go wrong.

As to how to achieve this: **Cameron and I need help understand where in the code we can inject extra logic at the end of the decommissioning process, before the transition to decommissioned**

Can one of you point us to where in the code that state is flipped, and what control component is responsible for it?

**Comment by irfansharif** [comment]
> Can one of you point us to where in the code that state is flipped, and what control component is responsible for it?

On the CLI side:

https://github.com/cockroachdb/cockroach/blob/e26432b8bc5a3f97e2ebedfa96745e2726d7ff11/pkg/cli/node.go#L434-L439

https://github.com/cockroachdb/cockroach/blob/e26432b8bc5a3f97e2ebedfa96745e2726d7ff11/pkg/cli/node.go#L470-L474

**Comment by knz** [comment]
Oh TIL - can you confirm that there's no other way for a cluster to mark a node as DECOMMISSIONED than in response to a CLI-issued RPC?

**Comment by knz** [comment]
(Irfan confirms)

**Comment by cameronnunez** [comment]
Can confirm in my initial stab at this I have been working where Irfan has pointed.

**Comment by tbg** [comment]
> Tobias and I don't have a preference about whether it should be at the beginning, during or at the end

I do have a preference actually. It shouldn't happen at the beginning. I think it should happen after the replicas have been drained, and before we flip DECOMMISSIONING->DECOMMISSIONED.

**Comment by bdarnell** [comment]
> Customer and Irfan say we want to drain SQL clients away at some point (instead of not at all, as currently implemented)
> Ben says this should happen at the end

You're reading more into my fault tolerance comments than I intended. I wasn't talking about draining SQL connections at all, but about what I understood to be a proposal to kill the node first and handle the replica movement as an after-the-fact repair.

My preference is to drain KV leases first, then SQL connections, then move all the replicas. Users should be encouraged to drain SQL connections via their own means before starting decommissioning; the drain on our side at step 2 would just be a stopgap for users who failed to do so.

> As a historical datapoint, we used to mark nodes as draining at the beginning of decommissioning, and this too produced lots of undesirable behavior for our customers who then tended to accidentally start decommissioning more nodes than they wanted, and tanked their clusters.

If they decommission too many nodes, aren't they out of luck no matter what we do here? How does draining at the end save them? Is it just that the replica movement won't succeed if there aren't enough surviving nodes to house them all? I think this argues for some sort of decommissioning lock/semaphore instead of delaying the SQL drain until the end. 


My argument for draining KV leases first is that it is internal to the DB and we can control the pace and process. If transferring our leases away causes the remaining nodes to get overloaded, we could (in principle, not in v1) detect this, reverse the process, and abort the decommission. Similarly, while draining SQL connections involves more factors outside our control, it's much cheaper than moving replicas, and if the decommissioning is going to cause problems better to find out at this stage rather than after going through the expensive replica moving process.

**Comment by tbg** [comment]
I'm fine with any solution here as long as we make sure a fat-finger is recoverable. Back in the day, draining was pretty aggressive (maybe it even stopped raft processing?) and so a user that started decommissioning a required set of nodes had trouble getting out of this state again - since it was hard to undo the decommissioning command since the cluster had gone unavailable, and since the draining status was sticky (i.e. nodes would start draining again after restart).
I'm probably leaning on these past experiences too much. Today's draining is much more well-rounded and in particular it's totally fine to have all nodes in the cluster draining (in the sense that KV stays healthy; of course all SQL ports will refuse connections and not all drain operations may be able to complete):

```
roachprod create -n 4 local
roachprod stage local release v21.2.2
roachprod start local
for i in 1 2 3; do roachprod run local:$i -- ./cockroach node drain --host=:{pgport:$i} --insecure& done
roachprod sql local:4
root@localhost:26263/defaultdb> create table foo(id int primary key, v string);
CREATE TABLE
```

So draining at the beginning seems ok as well. The purist in me dislikes how we bundle a drain operation into the decommissioning operation but if we are deciding that this is what users (SRE and on-prem) want, then so be it.

**Comment by tbg** [comment]
> however, a user who wants to guarantee a graceful drain at the end cannot do so today.

The question is if the user who wants that a) exists and b) is one we want to accommodate.

But, they also can, sort of, do this already, though it's annoying. You have to use `decommission --wait=none` and then... I don't know, wait for the replica count to drop to zero, then issue the drain, then re-invoke `decommission` to finish the job. With a new `--wait` mode that waits but doesn't flip to the final status, they could do it properly.

My understanding is that users just don't consistently drain their nodes before decommissioning them, and that they are running into problems as a result. So any way to bake the drain into the decommissioning at any appropriate time should solve the problem. Even a flag on decommission that requires explicit opt-in to decommission a live draining node should "solve" that particular problem (because users forgetting to drain will be reminded).

I do agree with an earlier commenter that this thread is too long. I am always operating under my own assumption of what the problem that we are solving actually is. A bit of synthesis here would be nice before a solution is being locked in. I do get the sense that most of the proposals here are trying to do way more than what is realistically needed, but I could be wrong.

**Comment by knz** [comment]
> The question is if the user who wants that a) exists and b) is one we want to accommodate.

The answer to this is yes to both. One of our major customers came to us (that is _exactly_ why this issue exists) and pointed out that their automation for decommissioning nodes broke in their upgrade to v21.x because previously they could assume that decommission implies drain, and now that's not true any more. They want us to fix that. 

> But, they also can, sort of, do this already, though it's annoying

Yes it has become annoying now, *and it was not annoying previously*, so we have a regression and we should fix it.

> My understanding is that users just don't consistently drain their nodes before decommissioning them, and that they are running into problems as a result.

That understanding may be correct (and warrant doc updates) but it doesn't detract from the fact we've created a problem that did not exist in pre-v21.x (= a regression), and we still need to fix it.

> Even a flag on decommission that requires explicit opt-in to decommission a live draining node should "solve" that particular problem.

That behavior, while it may be desirable in the future, would be a backward-incompatible change. If we introduce this, we need an intermediate version  that informs customers it will change. *Until then* we must still repair the regression that we introduced recently.

> A bit of synthesis here would be nice before a solution is being locked in

Synthesis is: **customer came to us and said we have a regression wrt SQL client behavior at the end of decommission, please fix it** and the way we should do that is re-introduce a proper graceful drain somewhere in the deco process. 

We can (and probably should) have a *separate* discussion about "what should be the ideal decommission journey" and how to document that, but for the customer at hand this story would be talking about the future, whereas htey want us to fix the *current* **regression** before we do anything else.

**Comment by tbg** [comment]
> because previously they could assume that decommission implies drain, and now that's not true any more. They want us to fix that.

That doesn't seem true, decommissioning hasn't implied draining since ~forever. What is new is that nodes get excluded from the cluster more aggressively. I bet this customer never drained their nodes and it just wasn't as obviously an issue before. They really should have done this for many releases (assuming our docs have been suggesting that, which I am also not sure is true). But either way, the regression is that not draining the node has become much worse than it used to be. We can't "repair" the regression without also introducing new behavior.

> Synthesis is: customer came to us and said we have a regression wrt SQL client behavior at the end of decommission, please fix it and the way we should do that is re-introduce a proper graceful drain somewhere in the deco process.

Again, I don't think draining was involved. What is new is that the node becomes unable to talk to the cluster and so SQL sessions hang. Previously, they would "just work" until the operator terminated the node (draining it or not in the process).

**Comment by knz** [comment]
> We can't "repair" the regression without also introducing new behavior.
> Previously, they would "just work" until the operator terminated the node (draining it or not in the process).

Don't you see the contradiction in these two sentences? Previously, the operator would see a decommissioned node appear to work as usual, and they could issue a graceful drain at the end, and that would work. Now they can't any more.

By adding an automatic graceful drain at the end (but not process termination), we restore something that looks & feels what the operators had previously. How is that "introducing new behavior"?

**Comment by tbg** [comment]
I don't want to quibble about semantics but in the synthesis you helpfully provided it comes across as we were previously draining the nodes as part of decommissioning. That is simply not correct and I thought that was worth pointing out. We cannot go back to not locking nodes out of the cluster at the end of decommissioning, so we must do something *else*, i.e. new functionality. Your suggestion to drain before flipping to decommissioned makes sense if the goal is to get as close as possible to the old behavior. It is also appealing to me from a practical point of view: draining a node that has no replicas is much less likely to fail or take a long time than draining at the beginning. =

As a heads up, we are about to land changes that let the leaseholder remove itself (https://github.com/cockroachdb/cockroach/pull/74077) so the point at which we drain really determines how data flows during decommissioning:

- if we drain first, then the remainder of the cluster sends all of the snapshots
- if we drain last, the node itself sends all of the snapshots for ranges for which it holds the leases.

The first option seems vaguely more appealing here since it should be faster overall. But it could also be more expensive in terms of transfer cost. Definitely if we drain the node early, we are working with fewer options in the allocator. So here too I would lean towards option two, draining as late as possible.

**Comment by knz** [comment]
It sounds like whether to drain early or late has an overall impact on the speed of rebalancing, *depending on the position of the decommissioned node in the multi-region setup*. That will be interesting to document.

That being said, this also highlight this is a *choice* that operators should be able to make, so we need to ensure they can in fact choose between the two options. As I explained above, the choice to drain early exists today. The choice to drain late, in contrast, doesn't. Adding an auto-drain at the end of decommissioning would fill this gap.

**Comment by bdarnell** [comment]
> As a heads up, we are about to land changes that let the leaseholder remove itself (#74077) so the point at which we drain really determines how data flows during decommissioning:

There are two kinds of draining, KV leases and SQL connections. It is the draining of KV leases that affects the data flow during decommissioning, while we're mainly talking about SQL connections here. (Unless I've really gotten lost).

**Comment by tbg** [comment]
That's only a subset of draining but yes, that would suffice, since the KV level "drains" by means of moving all replicas off the node.

**Comment by knz** [comment]
> There are two kinds of draining, KV leases and SQL connections. It is the draining of KV leases that affects the data flow during decommissioning, while we're mainly talking about SQL connections here.

Yes, the fact there are two kinds of draining has been discussed before. However, both @mwang1026 and I are not keen to separate them in the implementation, so as we can continue abstracting them as a single "drain" concept to end-users.

**Comment by bdarnell** [comment]
> Don't you see the contradiction between these two sentences? Previously, the operator would see a decommissioned node appear to work as usual, and they could issue a graceful drain at the end, and that would work. Now they can't any more.

Not every change in observable behavior is a regression. The question is whether this is something that they could have reasonably relied on before or if they got lucky with an implementation detail. I think the answer is that this is reasonable - even if it's not what we consider to be the ideal workflow, decommission should imply a drain at some point in the process. As I said earlier, I prefer draining early instead of late, but this is not a very strong preference and we should just pick one and go with it. 

> That being said, this also highlight this is a choice that operators should be able to make, so we need to ensure they can in fact choose between the two options.

Do we really need this? The customer in this issue is not saying "I can drain early but I'd rather drain late", they're saying "I don't want to have to think about draining at all". This is something we could consider building if there's demand for it, but it's a separate issue and I'd prefer we explore other ways of doing this more automatically first (like the ability to send snapshots from followers).

**Comment by knz** [comment]
> This is something we could consider building if there's demand for it.

Can you clarify what “something” refers to in this sentence?

**Comment by knz** [comment]
just had a quick sync with @jtsiros @bdarnell @tbg @mwang1026 
- prior behavior didn't have SQL availability gap between decommissioning and process termination; new behavior has such a gap, and that's the regression worth addressing.
- adding an auto-drain at the end of decommissioning (emphasis: -ing) before flip to decommisssioned will bridge that gap. This is what we need to change in the product now to help the customer.
- operators remain able to issue a drain prior to decommissioning (drain is idempotent).
- our docs on decommissioning and restart UX / best-practices are somewhat outdated, and will need to be improved. Separate project.

**Comment by tbg** [comment]
Reopening this to highlight that we got an escalation on 22.1.15 which includes #74319. The customer was getting routed to the decommissioned node and experienced non-retriable errors there. As I observed in a repro[^1], the likely reason is that while we start draining the node just before yanking it from the cluster, we don't actually wait for that to have any effect. So client traffic is affected, and since the node is immediately unable to contact its former peers, draining will not be very successful - it's likely going to take about as long as it can.

This was on cloud, btw, so likely the load balancers *did* remove the node pretty soon, but not soon enough for it to have an effect that prompted the customer to open an escalation which now hit the L2 rotation.

We should revisit if this is actually working well enough or not. I'm pretty sure if we ran a roachtest where a client targets the node that is being decommissioned, the client would get stuck in limbo for a while, instead of being "cleanly" refused by the SQL subsystem having fully drained.

[^1]: https://github.com/cockroachlabs/support/issues/2135#issuecomment-1453639169
